### PR TITLE
`Integrated code lifecycle`: Allow admins to set build timeout options via application properties

### PR DESCRIPTION
**Pull request by BBesrour**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

This PR allows admins to adjust the timeout options in the UI.
New properties are added:
- `artemis.continuous-integration.build-timeout-seconds.min`: Min timeout option
- `artemis.continuous-integration.build-timeout-seconds.max`: Max timeout option (if the user somehow sends a value higher than this, it will be checked and overridden)
- `artemis.continuous-integration.build-timeout-seconds.default`: default option

### Steps for Testing

Prerequisites:
- 1 Instructor
- Test server with ICL

1. Go to exercise creation, open dev tools
2. Check `Customize build script`
3. Make sure the default value of the timeout is 120 (default value), and that the max and min are respectively 240 and 10.
4. Save exercise
5. In the dev tools, check that the timeout was set correctly

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review

- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Performance Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| programming-exercise-build-configuration.component.ts | 95% | ✅ ❌ |
| profile-info.model.ts | 100% | ✅ |
| profile.service.ts | 92.94% | ✅ ❌ |

## Release Notes

- **New Features**
	- Enhanced build job management with increased timeout settings (default updated to 240 seconds).
	- Added configurable timeout options for instructor builds, including minimum, maximum, and default values.
	- Dynamic binding for timeout settings in the programming exercise build configuration component.
	- Expanded profile information to include build timeout settings for better management.
	- Introduced a new section in documentation for configuring build timeout settings.
	- Restructured user documentation for programming exercise setup, including a new section on editing maximum build duration.

- **Bug Fixes**
	- Improved error handling during build job execution, particularly for cancellations.

- **Tests**
	- Enhanced test coverage for the programming exercise build configuration component, validating timeout initialization logic based on profile information.

### CODE REVIEW

**Review by HawKhiem — COMMENTED** [review]
I tested on TS4. The min and default timeout is correct as described. The max timeout is however 240.

Screenshot 2024-10-27 003548
Screenshot 2024-10-27 003555
Screenshot 2024-10-27 003633

**Review by Cathy0123456789 — COMMENTED** [review]
Tested on TS4. Min and default timeout are set correctly, but max timeout is only 240s:
image

Not quite sure where to check step 5 of testing: "In the dev tools, check that the timeout was set correctly"?

**Comment by BBesrour** [comment]
@Cathy0123456789 @HawKhiem 240 as max value is correct. I mistakenly put 360 in the description (fixed it)

@Cathy0123456789 Press F12 in your browser and go to the network tab. Look for the request responsible for creating the exercise (endpoint should end with `/setup`) and then check the body (under buildConfig -> timeout)

**Review by SimonEntholzer — DISMISSED** [review]
Code

**Review by Cathy0123456789 — APPROVED** [review]
@BBesrour Thank you for the explanation!
I tested again on TS2, seems to be working good!

**Review by flbrgit — APPROVED** [review]
Tested on TS4, timout min max correct, timeout set correctly in responses, everything works as expected.
Screenshot 2024-10-28 122913

**Review by zagemello — APPROVED** [review]
image
All works as expected, tested on TS4, however on TS6 I was unable to find the option for the timeout :)

**Review by sarpsahinalp — DISMISSED** [review]
Code lgtm

**Review by Cathy0123456789 — APPROVED** [review]
Tested on TS1. Works as expected and timeout set correctly 
image

**Review by SimonEntholzer — DISMISSED** [review]
Re-approve after added docs

**Review by HawKhiem — APPROVED** [review]
Tested on TS3. Reapprove

image

File: docs/admin/setup/distributed.rst [hunk 1]
```diff
@@ -620,6 +619,8 @@ These credentials are used to clone repositories via HTTPS. You must also add th
                 build-agent:
                     short-name: "artemis-build-agent-X"             # Short name of the build agent. This should be unique for each build agent. Only lowercase letters, numbers and hyphens are allowed.
                     display-name: "Artemis Build Agent X"           # This value is optional. If omitted, the short name will be used as display name. Display name of the build agent. This is shown in the Artemis UI.
+                build-timeout-seconds:
+                    max: 2400                                      # (Optional, default 240) Maximum time in seconds a build job is allowed to run. If a build job exceeds this time, it will be cancelled.
```
Comment (JohannesStoehr) [inline docs/admin/setup/distributed.rst:623]: Why does this value deviate from the default?

File: src/main/resources/config/application-buildagent.yml [hunk 2]
```diff
@@ -34,6 +33,10 @@ artemis:
             expiry-minutes: 5
             cleanup-schedule-minutes: 60
         pause-grace-period-seconds: 60
+        build-timeout-seconds:
+          min: 10
+          default: 120
+          max: 240
```
Comment (JohannesStoehr) [inline src/main/resources/config/application-buildagent.yml:39]: These values do not align with the value mentioned in distributed.rst

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by HawKhiem — APPROVED** [review]
LGTM, reapprove

**Review by SimonEntholzer — APPROVED** [review]
reapprove

**Review by Cathy0123456789 — APPROVED** [review]
Reapprove

**Review by SindiBuklaji — APPROVED** [review]
Tested on Ts3 and everything worked as described 👍

**Review by kevinfischer4 — APPROVED** [review]
Tested on TS2. The timeout can be set to a value between 10 and 240. Works as expected.

**Review by sawys777 — APPROVED** [review]
Tested on TS1. Everything worked as expected and the range of timer was correct as well.

**Pull request merged**